### PR TITLE
Segmentations integration

### PR DESCRIPTION
**Pull request by jcfr**

### CODE REVIEW

File: Libs/vtkSegmentationCore/vtkTopologicalHierarchy.cxx [hunk 1]
```diff
@@ -185,6 +185,7 @@ void vtkTopologicalHierarchy::Update()
           if ((*it) == polyInIndex)
             {
             isContained = true;
+            break;
```
Comment (jcfr) [inline Libs/vtkSegmentationCore/vtkTopologicalHierarchy.cxx]: @cpinter Let me know what you think of this one.

**Comment by jcfr** [comment]
I am not sure yet how to address  warnings like these ones:

```
/home/jcfr/Projects/Slicer/Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:168: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
           if (*modifierImagePtr > maskThreshold)
                                 ^
```

and 

```
/home/jcfr/Projects/Slicer/Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx:123: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
           if (*modifierImagePtr > *baseImagePtr)
                                 ^
```

`maskThreshold` is an `int`. It could be templated over the type of the image to threshold ?

File: Libs/vtkSegmentationCore/vtkTopologicalHierarchy.cxx (hunk 1 above)
Comment (cpinter) [inline Libs/vtkSegmentationCore/vtkTopologicalHierarchy.cxx]: :+1:

**Comment by jcfr** [comment]
Similarly to what is done in [Imaging/Core/vtkImageThreshold.h](https://github.com/Kitware/VTK/blob/master/Imaging/Core/vtkImageThreshold.h), I suggest to have `maskThreshold` be a `double`.

Additionally, threshold should probably be updated to match the input type. May something like  [this](https://github.com/Kitware/VTK/blob/2c7c07ce63b1d29354a1674a0b947d38e230599f/Imaging/Core/vtkImageThreshold.cxx#L146-L180)

**Comment by cpinter** [comment]
I'd also refer this to @lassoan, as he added threshold to this function in SlicerRT rev2922.

**Comment by jcfr** [comment]
Et voila. I am done reviewing. 

After addressing the remaining warnings, we should be in good shape.

**Comment by cpinter** [comment]
Thanks a lot, Jc! As I said there will be one test failing (py_SegmentationTest1), and if it's alright I'll fix it right after integration, when I don't have to deal with nasty merging. Same goes for SlicerRT. I of course made the changes necessary to keep it working with the integrated Segmentations, but will test again and commit to SlicerRT once this pull request is integrated. Thanks again!

File: Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt [hunk 2]
```diff
@@ -77,6 +77,7 @@ set(${KIT}_TARGET_LIBRARIES
 SlicerMacroBuildModuleQtLibrary(
   NAME ${KIT}
   EXPORT_DIRECTIVE ${${KIT}_EXPORT_DIRECTIVE}
+  FOLDER "Module-${MODULE_NAME}"
```
Comment (cpinter) [inline Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt]: What does this do?

File: Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt [hunk 3]
```diff
@@ -62,8 +62,6 @@ set(${KIT}_RESOURCES
   Resources/${KIT}.qrc
   )
 
-SET (${KIT}_INCLUDE_DIRS ${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR} CACHE INTERNAL "" FORCE)
```
Comment (cpinter) [inline Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt]: Does this work outside Slicer core? Should I change these in SlicerRT?

**Comment by jcfr** [comment]
Commit c723ba6 integrated as [r25205](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25205)

File: Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt (hunk 2 above)
Comment (jcfr) [inline Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt]: This allow to keep targets organized in Visual Studio.

File: Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt (hunk 3 above)
Comment (jcfr) [inline Modules/Loadable/Segmentations/EditorEffects/CMakeLists.txt]: I am having an other look .. clean build us failing.

**Comment by cpinter** [comment]
@jcfr It seems to build now. What's next?

**Comment by jcfr** [comment]
Indeed, this is great.

Before integrating, I will plan on:
* disabling test `py_SegmentationsModuleTest1` (it failed on my workstation)
* entering a Slicer issue to track the warning related to `-Wsign-compare` discussed above

If that sound reasonable to you, I will merge this afternoon.

**Comment by cpinter** [comment]
Yes, sounds great!
As I said before, py_SegmentationsModuleTest1 is supposed to fail for now (RT data is loaded into it and I need to change it to non-DICOM files, that's all). I'll fix that and re-enable shortly after integration.
Thank you!

**Comment by jcfr** [comment]
Great.

Do you have the list of all individuals who co-authored the topic ? 

So far, I have:

```
From: Csaba Pinter <csaba.pinter@queensu.ca>

Co-authored-by: 
  Andras Lasso <lasso@queensu.ca>
  Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>  (build infrastructure and compilation warning fixes)
  
```

**Comment by cpinter** [comment]
Kyle Sunderland <1krs1@queensu.ca> (developed vtkPlanarContourToClosedSurfaceConversionRule)
Adam Rankin <arankin@robarts.ca> (initial implementation round for segmentation MRML and containers)

**Comment by jcfr** [comment]
@cpinter I squashed the commit and updated the commit message. 

There are also two more tweaks I am considering, let me know what you think.

Note we can also remove `vtkSegmentationCoreConfig.cmake` at some later point ... I am just wondering what was the use case.

**Comment by cpinter** [comment]
The use case for vtkSegmentationCoreConfig.cmake was probaby to allow using it from a superbuild. It was added very long ago (Dec 2014, rev2075). I think it can be removed

**Comment by jcfr** [comment]
Integrated in [25206](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=25206)

Three issues have also been created:
* [#4230](http://na-mic.org/Mantis/view.php?id=4230) Segmentations: Remove vtkImageFillROI
* [#4229](http://na-mic.org/Mantis/view.php?id=4229) Segmentations: vtkOrientedImageDataResample - Fix sign compare warnings
* [#4228](http://na-mic.org/Mantis/view.php?id=4228) Segmentations: Fix SegmentationsModuleTest1

**Comment by cpinter** [comment]
Thank you very much for working on this, Jc!

**Pull request closed without merging**